### PR TITLE
Expanded skip_file config parameter to support spaces and pathing

### PR DESCRIPTION
**Pull request by robertjf**
expanded the wild2regex conversion method to handle directory separators and spaces to allow for config like the following:

```
skip_file = ".*|~*|Desktop|Documents/OneNote*|Documents/IISExpress|Documents/SQL Server Management Studio|Documents/Visual Studio*|Documents/config.xlaunch|Documents/WindowsPowerShell"
```

Have also updated isPathExcluded to check the regex as well.

Based on the example above, this results in the directory `Documents/OneNote Notebooks` being excluded for example.

Also added a little verbose logging to validate this, which could probably be removed or cleaned up.

### CODE REVIEW

File: src/selective.d [hunk 1]
```diff
@@ -33,13 +34,17 @@ final class SelectiveSync
 		// Does the file match skip_file config entry?
 		// Returns true if the file matches a skip_file config entry
 		// Returns false if no match
+		log.vlog("isNameExcluded for name '", name, "': ", !name.matchFirst(mask).empty);
```
Comment (abraunegg) [inline src/selective.d]: I like the idea of logging if something is matched, however at a verbose level this becomes too much - potentially better suited to a --debug flag option at runtime

File: src/selective.d [hunk 2]
```diff
@@ -33,13 +34,17 @@ final class SelectiveSync
...
 		return !name.matchFirst(mask).empty;
 	}
 
 	// config sync_list file handling
+	// also incorporates skip_file config parameter for expanded regex path matching
 	bool isPathExcluded(string path)
 	{
-		return .isPathExcluded(path, paths);
+		log.vlog("isPathExcluded for path '", path, "': ", .isPathExcluded(path, paths));
```
Comment (abraunegg) [inline src/selective.d]: Same comment here for these two verbose outputs

**Comment by abraunegg** [comment]
I like the idea of logging if something is matched, however at a verbose level this becomes too much - potentially better suited to a --debug flag option at runtime.

Can you update your patch to include a new --debug flag with a new 'log.debug' function (or something along those lines) so that not only that function can be leveraged elsewhere, but also for items like this were we can easily see what is being matched & potentially see why something is failing by increasing the logging from verbose to debug.

When testing the patch 'as-is', using a sample directory `Documents/OneNote Notebooks` I can't seem to get your patch to skip the directory as required:

```
[root@localhost onedrive]# cat ~/.config/onedrive/config 
skip_file = "~*|Desktop|Documents/OneNote*|Documents/IISExpress|Documents/SQL Server Management Studio|Documents/Visual Studio*|Documents/config.xlaunch|Documents/WindowsPowerShell"
[root@localhost onedrive]# 
```

```
Loading config ...
Using Config Dir: /root/.config/onedrive
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /root/OneDrive
Initializing the Synchronization Engine ...
Account Type: business
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 1099364781983
Fetching details for OneDrive Root
OneDrive Root exists in the database
Applying changes of Path ID: <redacted>
Uploading differences of .
Processing root
isNameExcluded for name 'root': false
isPathExcluded for path '.': false
Path Matched for path '.': false
The directory has not changed
Uploading new items of .
isNameExcluded for name 'Documents': false
isPathExcluded for path './Documents': false
Path Matched for path './Documents': false
OneDrive Client requested to create remote path: ./Documents
The requested directory to create was not found on OneDrive - creating remote directory: ./Documents
Successfully created the remote directory ./Documents on OneDrive
isNameExcluded for name 'OneNote Notebooks': false
isPathExcluded for path './Documents/OneNote Notebooks': false
Path Matched for path './Documents/OneNote Notebooks': false
OneDrive Client requested to create remote path: ./Documents/OneNote Notebooks
The requested directory to create was not found on OneDrive - creating remote directory: ./Documents/OneNote Notebooks
Successfully created the remote directory ./Documents/OneNote Notebooks on OneDrive
isNameExcluded for name 'test.txt': false
isPathExcluded for path './Documents/OneNote Notebooks/test.txt': false
Path Matched for path './Documents/OneNote Notebooks/test.txt': false
Uploading file ./Documents/OneNote Notebooks/test.txt ...
Uploading fragment: 0-9/9
 done.
Remaining free space: 1099364781974
Applying changes of Path ID: <redacted>
isNameExcluded for name 'Documents': false
isPathExcluded for path 'Documents': false
Path Matched for path 'Documents': false
isNameExcluded for name 'OneNote Notebooks': false
isPathExcluded for path 'Documents/OneNote Notebooks': false
Path Matched for path 'Documents/OneNote Notebooks': true
isNameExcluded for name 'test.txt': false
```

My understanding of your patch is that `/Documents/OneNote Notebooks` should be skipped - the directory 'may' be created, however the test.txt should not be uploaded

**Comment by robertjf** [comment]
Good idea re the debug log level - have implemented that; although I'm wondering whether we should have a combined log level e.g. `--log=INFO|VERBOSE|DEBUG` or something like that.  I've updated the README.md file with the additional switch.  Also, currently --debug doesn't inherently enable --verbose, which it probably should.

Regarding creating the OneNote Notebook directory - locally my patch is ignoring it altogether and it doesn't get created:

```
[DEBUG] isNameExcluded for name 'OneNote Notebooks': false
[DEBUG] isPathExcluded for path 'Documents/OneNote Notebooks': false
[DEBUG] Path Matched for path 'Documents/OneNote Notebooks': true
```

if it was being created you'd see something like this:

```
[DEBUG] isNameExcluded for name 'Software': false
[DEBUG] isPathExcluded for path 'Documents/Software': false
[DEBUG] Path Matched for path 'Documents/Software': false
Creating directory: Documents/Software
```

Not sure what you mean about the test.txt file - which directory will it be in?  Any files in the excluded directories are not being downloaded for me...

**Comment by abraunegg** [comment]
> Good idea re the debug log level - have implemented that; although I'm wondering whether we should have a combined log level e.g. --log=INFO|VERBOSE|DEBUG or something like that. 

The whole logging & output of various items are on the whiteboard to 'fix' so that a single switch is needed to increase the verbosity. Right now my action plan is to make sure the 'main' bugs and issues are closed out - with #23 being the last 'major' one that I have found.

> Not sure what you mean about the test.txt file - which directory will it be in? Any files in the excluded directories are not being downloaded for me...

Will check your PR again - as when I tested it earlier everything was still being uploaded to OneDrive.

```
mkdir -p "Documents/OneNote Notebooks"  
vi Documents/OneNote\ Notebooks/test.txt
```

Based on my understanding - 'Documents/OneNote Notebooks' should be skipped .. but it was not being skipped ..

**Comment by robertjf** [comment]
Ah, yes I see now - latest commit resolves this issue by recursively checking a path's components for a match against the regex generated from the `skip_file` config parameter.

I've also introduced some more debug logging and made the output a little more consistent

**Comment by abraunegg** [comment]
Thanks for this. I think what I will go is cherry pick the 'bug fix' part as a separate change to close out the bug you found, then keep this PR for the logging changes / enhancements. I want to get the bug fix included in the pending release that I have in draft, so that the only remaining bug is the quickxorhash one.

**Comment by robertjf** [comment]
sounds good; once you've done that I'll clear the pr and start using branches instead of master.

**Comment by robertjf** [comment]
closing this PR as the relevant changes have already been pulled in; debug logging is the only thing left here and needs to be considered as part of a larger piece of work

**Pull request closed without merging**